civis/service_client.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import absolute_import
+
+from collections import OrderedDict
+import re
+import warnings
+
+from jsonref import JsonRef
+import requests
+import six
+
+from civis import APIClient
+from civis.base import Endpoint, CivisAPIError, tostr_urljoin
+from civis.resources._resources import parse_method
+from civis._utils import to_camelcase
+
+
+def auth_service_session(session, service_id):
+    try:
+        client = APIClient()
+        service = client.services.get(service_id)
+    except CivisAPIError as err:
+        msg = ('There was an issue '
+               'finding service with ID {}.').format(service_id)
+        six.raise_from(ValueError(msg), err)
+
+    auth_url = service['current_deployment']['displayUrl']
+    # Make request for adding Authentication Cookie to session
+    session.get(auth_url)
+
+
+class ServiceEndpoint(Endpoint):
+
+    def __init__(self, session_kwargs, client,
+                 return_type='civis'):
+        self._session_kwargs = session_kwargs
+        self._return_type = return_type
+        self._client = client
+
+    def _build_path(self, path):
+        if not path:
+            return self._client._base_url
+        if not self._client._root_path:
+            return tostr_urljoin(self._client._base_url, path.strip("/"))
+        return tostr_urljoin(self._client._base_url,
+                             self._client._root_path.strip("/"),
+                             path.strip("/"))
+
+    def _make_request(self, method, path=None, params=None, data=None,
+                      **kwargs):
+        url = self._build_path(path)
+
+        with requests.Session() as sess:
+            auth_service_session(sess, self._client._service_id)
```

Comment:
Does this mean that every request from the generated client needs to be re-authorized? Or does `requests` automatically cache for a session?

Reply:
no, unfortunately it has to be reauthorized. when you retrieve a service from the api client, it has a url that will generate a session cookie, but the cookie expires after a few minutes. So for every request we are refreshing that url and hitting it first before hitting the actual request endpoint.